Add preamble and example to pattern matching in definitions (#2959)
ocaml/ocaml.org: data/tutorials/language/0it_00_values_functions.md

In both examples, `d` and `e` are local definitions.

> ## Introduction to Pattern Matching

Reviewer: ```suggestion
## Forms of Pattern Matching
```

> [Pattern matching](https://en.wikipedia.org/wiki/Pattern_matching) is a mechanism in programming language design that enables the destructuring of values and facilitates case analysis in a concise and expressive manner.

Reviewer: ```suggestion
[Pattern matching](https://en.wikipedia.org/wiki/Pattern_matching) is a programming language construct that generalizes case analysis, it makes subexpression inspection possible and applies to values of any type. 
```

> In the following sections, we explore matching inside `let` bindings. In the next chapter on [Basic Data Types and Pattern Matching](/docs/basic-data-types), we examine pattern matching in the construction of conditional expressions using `match...with` and compare it to `if...then...else`. In the chapter on [Error Handling](/docs/error-handling), we explore how destructuring values aids in error handling when using `try...with`.

Reviewer: ```suggestion
In the following sections, we explore matching inside `let` bindings, which is a special case. In the next chapter on [Basic Data Types and Pattern Matching](/docs/basic-data-types), we examine pattern matching in general cases using `match...with`. The chapter on [Error Handling](/docs/error-handling) explores how destructuring values aids in error handling when using `try...with`.
```

> In the following sections, we explore definitional pattern matching using `let` bindings. In the next chapter on [Basic Data Types and Pattern Matching](/docs/basic-data-types), we examine pattern matching in the construction of conditional expressions using `match...with` and compare it to `if...then...else`. In the chapter on [Error Handling](/docs/error-handling), we explore how destructuring values aids in error handling when using `try...with`.

Reviewer: ```suggestion
In the following sections, we explore matching inside `let` bindings. In the next chapter on [Basic Data Types and Pattern Matching](/docs/basic-data-types), we examine pattern matching in the construction of conditional expressions using `match...with` and compare it to `if...then...else`. In the chapter on [Error Handling](/docs/error-handling), we explore how destructuring values aids in error handling when using `try...with`.
```

val surname : string = "Milner"
```

### Nested Pattern Matching on User-Defined Types

Pattern matching also works with nested user-defined types.

> In the example below, we deconstruct nested tuples:

Reviewer: ```suggestion
Pattern matching also works with nested user-defined types. In the example below, we deconstruct nested tuples:
```

Avoid single sentence paragraphs

> ```ocaml

Reviewer 2: Generally it's a good idea to leave a blank line between code blocks and the surrounding text.

Reviewer 2: (even if the code in the guides doesn't do this consistently)

# let (name, (street, city, zip), (email, phone)) =
  ("John Doe", ("123 Elm St", "Springfield", 12345), ("john@example.com", 1234567890));;
val name : string = "John Doe"
val street : string = "123 Elm St"
val city : string = "Springfield"
val zip : int = 12345
val email : string = "john@example.com"
val phone : int = 1234567890
```
In the following example, we deconstruct nested records and an associated tuple.

> First, define an address record type

Reviewer 2: Probably such sentences should end with `:`.

```ocaml
# type address = {
  street: string;
  city: string;

>   zip: int

Reviewer: ```suggestion
  street : string;
  city : string;
  zip : int
```

  };;
type address = { street : string; city : string; zip : int; }
```

Define a person record type, containing the nested address record type
```ocaml
# type person = {
  name : string;
  address : address;  (* Reference the address type *)
  contact: string * int;
  };;
type person = { name : string; address : address; contact : string * int; }
```

Create an instance of the person type interface
```ocaml
# let john = {
  name = "John Doe";
  address = { street = "123 Elm St"; city = "Springfield"; zip = 12345 };
  contact = ("john@example.com", 1234567890);
};;
val john : person =
  {name = "John Doe";
   address = {street = "123 Elm St"; city = "Springfield"; zip = 12345};
   contact = ("john@example.com", 1234567890)}
```

Deconstructure the nested structure
```ocaml
# let { name; address = { street; city; zip }; contact = (email, phone) } = john;;
val name : string = "John Doe"
val street : string = "123 Elm St"
val city : string = "Springfield"
val zip : int = 12345
val email : string = "john@example.com"
val phone : int = 1234567890
```

> **note**: `address` and `contact` are not defined as variables available in your scope. Instead, they serve as binding patterns that help destructure the nested structure.

Reviewer 2: note -> Note, to be consistent with the rest of the docs.

Now, we can use `name`, `street`, `city`, `zip`, `email`, and `phone` as bindings
```ocaml
# let () =
  Printf.printf "Name: %s\nStreet: %s\nCity: %s\nZip: %d\nEmail: %s\nPhone: %d\n"

>     name street city zip email phone;;

Reviewer: ```suggestion
# Printf.printf "Name: %s\nStreet: %s\nCity: %s\nZip: %d\nEmail: %s\nPhone: %d\n"
    name street city zip email phone;;
```

`let () =` isn't needed in interactive top level

val surname : string = "Milner"
```

### Nested Pattern Matching on User-Defined Types

Pattern matching also works with nested user-defined types. In the example below, we deconstruct nested tuples:

```ocaml
# let (name, (street, city, zip), (email, phone)) =
  ("John Doe", ("123 Elm St", "Springfield", 12345), ("john@example.com", 1234567890));;
val name : string = "John Doe"
val street : string = "123 Elm St"
val city : string = "Springfield"
val zip : int = 12345
val email : string = "john@example.com"
val phone : int = 1234567890
```

In the following example, we deconstruct a tuple nested within a record.

> Records require explicit type definitions wherein field names are annotated with field types. We first define a `person` record type, then create an instance of that type named `jane`:

Reviewer: ```suggestion
Records require explicit type definitions wherein field names are annotated with types. We first define a `person` record type, then create an instance of that type named `jane`:
```

```ocaml
# type person = {
    name : string;
    street : string;
    city : string;
    zip : int;
    contact: string * int;
  };;
type person = {
  name : string;
  street : string;
  city : string;
  zip : int;
  contact : string * int;

# let jane = {
    name = "Jane Doe";
    street = "123 Elm St";
    city = "Springfield";
    zip = 12345;
    contact = ("jane@example.com", 1234567890);
  };;
val jane : person =
  {name = "Jane Doe"; street = "123 Elm St"; city = "Springfield"; zip = 12345;
   contact = ("jane@example.com", 1234567890)}
```

> The following examples demonstrate two methods by which we can extract Jane's `email` and `phone` number data in the nested `contact` tuple. We will do so first by using nested deconstruction, then demonstrate a brute force approach by first extracting `content` followed by accessing the `email` and `phone` by deconstructing `contact`. <!-- Note: This sequence is pedagologically backwards perhaps, but the goal is to show that `contact` does not become a bound variable when using nested deconstruction. This is difficult to demonstrate in a single utop session if `contact` is bound in the brute force approach first, as it will not show the desired error message -->  

Reviewer: ```suggestion
The following examples demonstrate two methods by which we can extract Jane's `email` and `phone` number data in the nested `contact` tuple. We will do so first by using nested deconstruction, then demonstrate another approach by first extracting `content` followed by accessing the `email` and `phone` by deconstructing `contact`. 
```
>  Note: This sequence is pedagologically backwards perhaps, but the goal is to show that `contact` does not become a bound variable when using nested deconstruction. This is difficult to demonstrate in a single utop session if `contact` is bound in the brute force approach first, as it will not show the desired error message

Nice suggestion, thanks

First, lets use nested deconstruction to access the contents of the `contact` tuple directly:

```ocaml
# let { name; street; city; zip; contact = (email, phone) } = jane;;
val name : string = "Jane Doe"
val street : string = "123 Elm St"
val city : string = "Springfield"
val zip : int = 12345
val email : string = "jane@example.com"
val phone : int = 1234567890
```

Notice that `contact` is not available in our top-level scope:

```ocaml
# contact;;
Error: Unbound value contact
```

> This is because "contact" is not a bound variable, rather it is an intermediate pattern for deconstructive pattern-matching.

Reviewer: ```suggestion
This is because "contact" is not a top-level definition; rather, it is a local definition from deconstructive pattern-matching.
```

> Next, we demonstrate the brute force approach for accessing `email` and `phone`. This brings `contact` into our top-level scope and requires two steps:

Reviewer: ```suggestion
Next, we demonstrate the two-phase approach for accessing `email` and `phone`. This brings `contact` into our top-level scope and requires two steps:
```

```ocaml
(* Step 1: deconstruct all record fields *)
# let { name; street; city; zip; contact } = jane;;
val name : string = "Jane Doe"
val street : string = "123 Elm St"
val city : string = "Springfield"
val zip : int = 12345
val contact : string * int = ("jane@example.com", 1234567890)

(* Step 2: deconstruct the tuple *)
# let ( email, phone ) = contact;;
val email : string = "jane@example.com"
val phone : int = 1234567890
```

Notice that `contact` is now available at the top-level as a bound variable:

```ocaml
# contact;;
- : string * int = ("jane@example.com", 1234567890)
```

### Discarding Values Using Pattern Matching

When pattern matching, it is possible to discard or ignore values that are not desired. The method by which this is done depends on the data structure being destructured.

Continuing from the above `jane` record example, we can simply omit the `zip` field in the pattern if we don't want to bind it to a variable:

```ocaml
# let { name; street; city; contact = (email, phone) } = john;;
val name : string = "Nohn Doe"
val street : string = "123 Elm St"
val city : string = "Springfield"
val email : string = "jane@example.com"
val phone : int = 1234567890
```

> Tuples behave differently from records because they require positional consistency. To discard the `email` value from the tuple of the `contact` field, we need to use the catch-all pattern (`_`):

Reviewer: ```suggestion
Tuples behave differently from records; contained data is anonymous, and its position is used to access it. To discard the `email` value from the tuple of the `contact` field, we need to use the catch-all pattern (`_`):
```
